rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+have the values omitted.
+
+In variants B, it would probably make sense to require exact number of values to be specified, or follow the usual function call syntax rules.
+
+Note that since records are first class objects, you can export or import a record through a module boundary in the usual way:
+
+```
+local X = require(path).X
+local r = new X(1, 2)
+```
+
+### Generic records
+
+> TODO: This needs some thought; e.g., do we support explicit specification of record arguments at construction time and what's the syntax for that? Do we need to do this in the first proposal? This likely makes packed records effectively impossible to support at runtime without a huge amount of work, does that matter?
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations as usual. The unification rules say that a table can unify with a record if the fields
+match, which makes records similar to sealed tables from the type checking perspective. (note, this is hand wavy on subtyping rules)
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+> Alternatively, is `export type Record = Record` too awkward?
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
+    return self.x + self.y
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Of course, the type checker also knows that the record type has the metatable with the inferred type of the record shape.
+
+> TODO: Does the type checker need to understand the internal structure of the shape so that type checking works across modules, or is simply modeling this
+> as a metatable sufficient?
+
+### Object modeling
+
+> TODO: records don't support implementation inheritance and why it's a good thing
+
+## Drawbacks
+
+Adding a new data type that is cross-cutting (across syntax, semantics/compiler, semantics/type checking, and runtime) results in added complexity.
+
+The rigidity of records may make some applications hesitate to adopt them; e.g. you can't simply add a new field at a random point in the program, which some would
+argue makes the language less dynamic and therefore less convenient.
+
+Not enforcing type compatibility for typed records at runtime may make it difficult for us to optimize record storage more by removing the type tags (which could
+make record objects ~2x more efficient in some cases).
+
```

Comment:
I don't think there's a problem with D v1 and D v2 in that sense. When you load the same module, you get the same record shape. If it updates, your code still works if it uses existing fields; obviously, removing a field that was accessed will trigger errors.

If you already load two different modules, you get two different shapes - yes! But the fact that these modules use records doesn't make anything harder or easier. Passing one record between B and C may or may not result in issues because the fields that it exposes and methods it offers may or may not have the same behavior. But that's exactly the same as with tables, so I don't think it's fair to say that this is a drawback of this proposal.

Python, Ruby, TypeScript are all examples of dynamic languages that do provide nominal types.

Reply:
The second drawback you listed is in Alternatives section.

Reply:
D v1 and D v2 are different modules though, it's just it's very easy for B to update D but not C, and as a result you get type errors.This is more an issue with the type system than the runtime.

You're right though that there are dynamic languages with nominal types, the nominal types don't support duck-typing though (at least not statically).

Reply:
Yeah I mean the semantics of the execution here, not the type feedback. For type safety it's interesting to ask if D v1 and D v2 are really compatible if the module is different and the shape is the same... there may be surprising results because the module state is actually distinct if the module comes from two different sources.

I'm not sure we're talking about the same thing. All languages I've specified behave as if duck typed at runtime. From the type checker perspective, there's leeway to do either here I think; I was assuming we'd want to go with nominal typing but we could certainly also go with structural.